File: micronaut-mcp-server-java-sdk/src/test/java/io/micronaut/mcp/server/stateless/sync/resources/ResourcesFactory.java
package io.micronaut.mcp.server.stateless.sync.resources;

/*
//tag::fakepackage[]
package example.micronaut;

//end::fakepackage[]
 */
//tag::imports[]
import io.micronaut.context.BeanContext;
import io.micronaut.context.annotation.Context;
import io.micronaut.context.annotation.EachBean;
import io.micronaut.context.annotation.Factory;
import io.micronaut.context.annotation.Requires;
import io.micronaut.context.exceptions.ConfigurationException;
import io.micronaut.core.io.ResourceLoader;
import io.micronaut.mcp.server.utils.PgnLoader;
import io.modelcontextprotocol.server.McpStatelessServerFeatures;
import io.modelcontextprotocol.spec.McpSchema;
import jakarta.inject.Singleton;

import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

//end::imports[]
@Requires(property = "spec.name", value = "StatelessSyncResourceListTest")
//tag::clazz[]
@Context
@Factory
class ResourcesFactory {
    public static final String PGN_MIME_TYPE = "application/x-chess-pgn";

    private final PgnLoader pgnLoader;
    private final ResourceLoader resourceLoader;

    ResourcesFactory(ResourceLoader resourceLoader, PgnLoader pgnLoader) {
        this.resourceLoader = resourceLoader;
        this.pgnLoader = pgnLoader;
    }

    @EachBean(PgnFile.class)
    @Singleton
    McpStatelessServerFeatures.SyncResourceSpecification createPgnSyncResourceSpecification(PgnFile pgnFile) {
        McpSchema.Resource resource = getResource(pgnFile);
        return new McpStatelessServerFeatures.SyncResourceSpecification(resource,
            (mcpSyncServerExchange,  readResourceRequest) -> readResourceResult(readResourceRequest.uri(), pgnLoader));
    }

    private static Integer round(String uri) {
        int lastSlash = uri.lastIndexOf('/');
        String roundStr = uri.substring(lastSlash + 1);
        return Integer.parseInt(roundStr);
    }

    static McpSchema.ReadResourceResult readResourceResult(String uri, PgnLoader pgnLoader) {
        Integer round = round(uri);
        List<McpSchema.ResourceContents> contents = new ArrayList<>();
        pgnLoader.loadPgn(round).ifPresent(text ->
            contents.add(new McpSchema.TextResourceContents(uri, PGN_MIME_TYPE, text)));
        return new McpSchema.ReadResourceResult(contents);
    }

    private McpSchema.Resource getResource(PgnFile pgnFile) {
        Optional<InputStream> roundPgnInputStreamOptional = resourceLoader.getResourceAsStream(pgnFile.getPath());
        if (roundPgnInputStreamOptional.isPresent()) {
            try {
                Integer round = pgnFile.getRound();
                Long size = Long.valueOf(roundPgnInputStreamOptional.get().readAllBytes().length);
                String uri = "pgn://round/" + round;
                String name = "round" + round + "PgnFideWCC2024";
                String title = "PGN of the Round " + round + " game of the World Chess Championship";
                String description = title + " between Ding Liren and Gukesh Dommaraju";
                return new McpSchema.Resource(uri, name, title, description, PGN_MIME_TYPE, size, null, null);
            } catch (IOException e) {
                throw new ConfigurationException("unable to calculate the size of the resource");
            }
        }
        throw new ConfigurationException("unable find resource for path " + pgnFile.getPath());
    }
}

File: test-suite-docs-python/src/test/python/micronaut/mcp/docs/resources/ResourcesFactory.py
# tag::imports[]
from io.modelcontextprotocol.server import McpStatelessServerFeatures
from io.modelcontextprotocol.spec import McpSchema
from jakarta.inject import Singleton
from java.util import ArrayList
from micronaut.context.annotation import Context, EachBean, Factory, Requires
from micronaut.context.exceptions import ConfigurationException
from micronaut.core.io import ResourceLoader

from .PgnFile import PgnFile
from .PgnLoader import PgnLoader
# end::imports[]


# tag::clazz[]
PGN_MIME_TYPE = "application/x-chess-pgn"


def round_of(uri: str) -> int:
    last_slash = uri.rfind("/")
    return int(uri[last_slash + 1:])


def read_resource_result(uri: str, pgn_loader: PgnLoader) -> McpSchema.ReadResourceResult:
    round = round_of(uri)
    contents = ArrayList()
    text = pgn_loader.load_pgn(round)
    if text is not None:
        contents.add(McpSchema.TextResourceContents(uri, PGN_MIME_TYPE, text))
    return McpSchema.ReadResourceResult(contents)


# end::clazz[]
@Requires(property="spec.name", value="ResourcesFactoryTest")
# tag::clazz[]
@Context
@Factory
class ResourcesFactory:

    def __init__(self, resource_loader: ResourceLoader, pgn_loader: PgnLoader):
        self.resource_loader = resource_loader
        self.pgn_loader = pgn_loader

    @EachBean(PgnFile)
    @Singleton
    def create_pgn_sync_resource_specification(self, pgn_file: PgnFile) -> McpStatelessServerFeatures.SyncResourceSpecification:
        resource = self.get_resource(pgn_file)
        return McpStatelessServerFeatures.SyncResourceSpecification(
            resource,
            lambda ctx, read_resource_request: read_resource_result(read_resource_request.uri(), self.pgn_loader))

    def get_resource(self, pgn_file: PgnFile) -> McpSchema.Resource:
        size = self.size(pgn_file.path)
        if size is None:
            raise ConfigurationException("unable find resource for path " + pgn_file.path)
        round = pgn_file.round
        uri = f"pgn://round/{round}"
        name = f"round{round}PgnFideWCC2024"
        title = f"PGN of the Round {round} game of the World Chess Championship"
        description = f"{title} between Ding Liren and Gukesh Dommaraju"
        return McpSchema.Resource(uri, name, title, description, PGN_MIME_TYPE, size, None, None)

    def size(self, path: str) -> int | None:
        input_stream_optional = self.resource_loader.getResourceAsStream(path)
        if input_stream_optional.isPresent():
            input_stream = input_stream_optional.get()
            try:
                return len(input_stream.readAllBytes())
            finally:
                input_stream.close()
        return None
# end::clazz[]
